consensus/core/src/authority_service.rs

```diff
@@ -356,13 +361,59 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
     }
 }
 
+/// Atomically counts the number of active subscriptions to the block broadcast stream,
+/// and dispatch commands to core based on the changes.
+struct SubscriptionCounter {
+    counter: parking_lot::Mutex<usize>,
+    dispatcher: Arc<dyn CoreThreadDispatcher>,
+}
+
+impl SubscriptionCounter {
+    fn new(dispatcher: Arc<dyn CoreThreadDispatcher>) -> Self {
+        Self {
+            counter: parking_lot::Mutex::new(0),
+            dispatcher,
+        }
+    }
+
+    fn increment(&self) -> Result<(), ConsensusError> {
+        let prev_subscriptions;
+        {
+            let mut counter = self.counter.lock();
+            prev_subscriptions = *counter;
+            *counter += 1;
```

Comment:
Yeah good reminder. I forgot to update the logic after making the call to dispatcher sync. The reason to use Mutex is to ensure atomic counter and core update. I wanted to avoid the following case: thread A decrements the counter to 0, thread B increments the counter to 1, thread B sets true, thread A sets false.